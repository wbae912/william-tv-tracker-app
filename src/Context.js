import React from 'react';

const tvContext = React.createContext({
  shows: [],
  deleteTvShow: () => {},
  addTvShow: () => {}
})

export default tvContext;