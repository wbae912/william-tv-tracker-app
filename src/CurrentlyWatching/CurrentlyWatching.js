import React, { Component } from 'react';
import TvShowEntry from '../TvShowEntry/TvShowEntry';
import tvContext from '../Context';
import './CurrentlyWatching.css';

export default class CurrentlyWatching extends Component {
  static contextType = tvContext;

  constructor(props) {
    super(props)
  
    this.state = {
       searchTerm: '',
       genre: '',
       minRating: '',
       sort: ''
    }
  }

  componentDidMount() {
    this.context.getAllShows();
  }

  handleSearchTermChange = (e) => {
    this.setState({
      searchTerm: e.target.value
    })
  }

  handleGenreChange = (e) => {
    this.setState({
      genre: e.target.value
    })
  }

  handleMinRatingChange = (e) => {
    this.setState({
      minRating: e.target.value
    })
  }

  handleSortChange = (e) => {
    this.setState({
      sort: e.target.value
    })
  }
  
  render() {
    let currentlyWatchingShows = this.context.shows.filter(show => show.status === 'Currently Watching');

    if(this.state.searchTerm) {
      currentlyWatchingShows = currentlyWatchingShows.filter(show => show.tv_title.toLowerCase().indexOf(this.state.searchTerm.toLowerCase()) !== -1);
    }
    if(this.state.genre) {
      currentlyWatchingShows = currentlyWatchingShows.filter(show => show.genre === this.state.genre);
    }
    if(this.state.minRating) {
      currentlyWatchingShows = currentlyWatchingShows.filter(show => show.rating >= this.state.minRating);
    }
    if(this.state.sort === '1') {
      currentlyWatchingShows = currentlyWatchingShows.sort((a,b) => (a.tv_title > b.tv_title) ? 1 : ((b.tv_title > a.tv_title) ? -1 : 0))
    }
    if(this.state.sort === '2') {
      currentlyWatchingShows = currentlyWatchingShows.sort((a,b) => (b.tv_title > a.tv_title) ? 1 : ((a.tv_title > b.tv_title) ? -1 : 0))
    }
    if(this.state.sort === '3') {
      currentlyWatchingShows = currentlyWatchingShows.sort((a,b) => a.rating - b.rating)
    }
    if(this.state.sort === '4') {
      currentlyWatchingShows = currentlyWatchingShows.sort((a,b) => b.rating - a.rating)
    }
    return (
      <div className="currently-section">
        <section className="tv-queries">
          <h1 className="current-h1">Currently Watching</h1>
          <div className="search-div">
            <label htmlFor="searchTerm" className="search-label">Search TV Show</label>
            <input type="text" name="searchTerm" id="searchTerm" placeholder="Search TV Show"
              value={this.state.searchTerm} 
              onChange={this.handleSearchTermChange} />
          </div>
          <div className="filter-genre">
            <select name="genre-options" id="genre-options" onChange={this.handleGenreChange}>
              <option value="">Filter By Genre...</option>
              <option value="N/A">N/A</option>
              <option value="Action">Action</option>
              <option value="Animated">Animated</option>
              <option value="Comedy">Comedy</option>
              <option value="Documentary">Documentary</option>
              <option value="Drama">Drama</option>
              <option value="Educational">Educational</option>
              <option value="Fantasy">Fantasy</option>
              <option value="Horror">Horror</option>
              <option value="Mystery">Horror</option>
              <option value="Reality">Reality</option>
              <option value="Sitcom">Sitcom</option>
              <option value="Sci-Fi">Sci-Fi</option>
              <option value="Thriller">Thriller</option>
              <option value="Variety">Variety</option>
            </select>
          </div>
          <div className="filter-rating">
            <select name="rating-options" id="rating-options" onChange={this.handleMinRatingChange}>
              <option value="">Filter By Min. Rating...</option>
              <option value="0">No Rating</option>
              <option value="1">1 Star</option>
              <option value="2">2 Stars</option>
              <option value="3">3 Stars</option>
              <option value="4">4 Stars</option>
              <option value="5">5 Stars</option>
            </select>
          </div>
          <div className="sort-div">
            <select name="sort-options" id="sort-options" onChange={this.handleSortChange}>
                <option value="0">Sort By...</option>
                <optgroup label="TV Show Name">
                  <option value="1">Name (A-Z)</option>
                  <option value="2">Name (Z-A)</option>
                </optgroup>
                <optgroup label="Rating">
                  <option value="3">Rating (Low to High)</option>
                  <option value="4">Rating (High to Low)</option>
                </optgroup>
            </select>
          </div>
        </section>

        {currentlyWatchingShows.map(show => 
          <TvShowEntry
            key={show.id}
            show={show}
            history={this.props.history}
          />)}

      </div>
    )
  }
}
