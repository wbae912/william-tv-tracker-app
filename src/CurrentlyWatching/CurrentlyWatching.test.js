import React from 'react';
import { shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import Context from '../Context';
import TokenService from '../services/token-service';
import CurrentlyWatching from './CurrentlyWatching';

describe(`CurrentlyWatching component`, () => {
  const props = {
    show:
      {
        "id": "1",
        "tv_title": "Example",
        "status": "Currently Watching",
        "season_number": "1",
        "episode_number": "1",
        "rating": "3",
        "genre": "Action",
        "description": "Lorem ipsum",
        "review": "Lorem ipsum"
      },
    key: "1"
  }

  const shows = [
      {
        "id": "1",
        "tv_title": "Example",
        "status": "Currently Watching",
        "season_number": "1",
        "episode_number": "1",
        "rating": "3",
        "genre": "Action",
        "description": "Lorem ipsum",
        "review": "Lorem ipsum"
      },
      {
        "id": "2",
        "tv_title": "Example 2",
        "status": "Currently Watching",
        "season_number": "1",
        "episode_number": "1",
        "rating": "3",
        "genre": "Action",
        "description": "Lorem ipsum",
        "review": "Lorem ipsum"
      }
    ];
  
  const getAllShows = () => {
    return fetch(`http://localhost:8000/api/shows/all`, {
      headers: {
      'authorization': `bearer ${TokenService.getAuthToken()}`
      }
    })
    .then(res => {
      if(!res.ok) {
        return res.json().then(err => Promise.reject(err));
      }
      return res.json();
    })
    .then(data => {
      this.setState({
        shows: data,
        isLoggedIn: true
      })
    })
  }
  
  it('renders component without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <Context.Provider value={{getAllShows, shows}}><CurrentlyWatching /></Context.Provider>
      </MemoryRouter>
      , div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders UI as expected', () => {
    const tree = renderer
      .create(<MemoryRouter><Context.Provider value={{getAllShows, shows}}><CurrentlyWatching /></Context.Provider></MemoryRouter>)
      .toJSON();
      expect(tree).toMatchSnapshot();
  })
  it('renders a .CurrentlyWatching by default', () => {
    const wrapper = shallow(<Context.Provider value={{shows}}><CurrentlyWatching /></Context.Provider>)
    expect(toJson(wrapper)).toMatchSnapshot()
  });
  it('renders a Show in section for each notes in array', () => {
    const section = shallow(<Context.Provider value={(shows)}><CurrentlyWatching {...props} /></Context.Provider>)
      .find('section')
    expect(toJson(section)).toMatchSnapshot()
  });
});