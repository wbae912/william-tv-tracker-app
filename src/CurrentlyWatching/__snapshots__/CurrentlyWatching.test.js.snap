// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentlyWatching component renders UI as expected 1`] = `
<div
  className="currently-section"
>
  <div
    className="error-bigger"
    role="alert"
  />
  <section
    className="tv-queries"
  >
    <h1
      className="current-h1"
    >
      Currently Watching
    </h1>
    <div
      className="query-flex"
    >
      <div
        className="search-div"
      >
        <label
          className="search-label"
          htmlFor="searchTerm"
        >
          Search TV Show
        </label>
        <input
          id="searchTerm"
          name="searchTerm"
          onChange={[Function]}
          placeholder="Search TV Show"
          type="text"
          value=""
        />
      </div>
      <div
        className="filter-genre"
      >
        <label
          aria-hidden="true"
          className="hidden"
          htmlFor="genre-options"
        >
          Genre Options
        </label>
        <select
          id="genre-options"
          name="genre-options"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Filter By Genre...
          </option>
          <option
            value="N/A"
          >
            N/A
          </option>
          <option
            value="Action"
          >
            Action
          </option>
          <option
            value="Animated"
          >
            Animated
          </option>
          <option
            value="Comedy"
          >
            Comedy
          </option>
          <option
            value="Documentary"
          >
            Documentary
          </option>
          <option
            value="Drama"
          >
            Drama
          </option>
          <option
            value="Educational"
          >
            Educational
          </option>
          <option
            value="Fantasy"
          >
            Fantasy
          </option>
          <option
            value="Horror"
          >
            Horror
          </option>
          <option
            value="Mystery"
          >
            Mystery
          </option>
          <option
            value="Reality"
          >
            Reality
          </option>
          <option
            value="Sitcom"
          >
            Sitcom
          </option>
          <option
            value="Sci-Fi"
          >
            Sci-Fi
          </option>
          <option
            value="Thriller"
          >
            Thriller
          </option>
          <option
            value="Variety"
          >
            Variety
          </option>
        </select>
      </div>
      <div
        className="filter-rating"
      >
        <label
          aria-hidden="true"
          className="hidden"
          htmlFor="rating-options"
        >
          Rating Options
        </label>
        <select
          id="rating-options"
          name="rating-options"
          onChange={[Function]}
        >
          <option
            value=""
          >
            Filter By Min. Rating...
          </option>
          <option
            value="0"
          >
            No Rating
          </option>
          <option
            value="1"
          >
            1 Star
          </option>
          <option
            value="2"
          >
            2 Stars
          </option>
          <option
            value="3"
          >
            3 Stars
          </option>
          <option
            value="4"
          >
            4 Stars
          </option>
          <option
            value="5"
          >
            5 Stars
          </option>
        </select>
      </div>
      <div
        className="sort-div"
      >
        <label
          aria-hidden="true"
          className="hidden"
          htmlFor="sort-options"
        >
          Sort Options
        </label>
        <select
          id="sort-options"
          name="sort-options"
          onChange={[Function]}
        >
          <option
            value="0"
          >
            Sort By...
          </option>
          <optgroup
            label="TV Show Name"
          >
            <option
              value="1"
            >
              Name (A-Z)
            </option>
            <option
              value="2"
            >
              Name (Z-A)
            </option>
          </optgroup>
          <optgroup
            label="Rating"
          >
            <option
              value="3"
            >
              Rating (Low to High)
            </option>
            <option
              value="4"
            >
              Rating (High to Low)
            </option>
          </optgroup>
        </select>
      </div>
    </div>
  </section>
  <ul
    aria-live="polite"
    className="tv-entries-list"
  >
    <li
      className="tv-show-entry"
      onClick={[Function]}
    >
      <h2
        className="entry-title"
      >
        Example
      </h2>
      <div
        className="entry-div"
      >
        <p
          className="entry-p"
        >
          Season: 
          1
        </p>
        <p
          className="entry-p"
        >
          Episode: 
          1
        </p>
        <p
          className="entry-p"
        >
          Genre: 
          Action
        </p>
        <p
          className="entry-p"
        >
          Rating: 
          3
        </p>
      </div>
      <div
        aria-live="polite"
        className="entry-buttons"
      >
        <i
          aria-label="Edit TV show button"
          className="fa fa-edit fa-lg edit-button"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        />
        <i
          aria-label="Delete TV show button"
          className="fa fa-trash fa-lg"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        />
      </div>
    </li>
    <li
      className="tv-show-entry"
      onClick={[Function]}
    >
      <h2
        className="entry-title"
      >
        Example 2
      </h2>
      <div
        className="entry-div"
      >
        <p
          className="entry-p"
        >
          Season: 
          1
        </p>
        <p
          className="entry-p"
        >
          Episode: 
          1
        </p>
        <p
          className="entry-p"
        >
          Genre: 
          Action
        </p>
        <p
          className="entry-p"
        >
          Rating: 
          3
        </p>
      </div>
      <div
        aria-live="polite"
        className="entry-buttons"
      >
        <i
          aria-label="Edit TV show button"
          className="fa fa-edit fa-lg edit-button"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        />
        <i
          aria-label="Delete TV show button"
          className="fa fa-trash fa-lg"
          onClick={[Function]}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
        />
      </div>
    </li>
  </ul>
</div>
`;

exports[`CurrentlyWatching component renders a .CurrentlyWatching by default 1`] = `<CurrentlyWatching />`;

exports[`CurrentlyWatching component renders a Show in section for each notes in array 1`] = `null`;
